File: thesarvo-android/guide/src/main/java/com/thesarvo/guide/GuideDownloader.java
package com.thesarvo.guide;


import android.content.Context;
import android.util.Log;

import com.google.common.collect.Iterables;

import org.apache.commons.io.FileUtils;
import org.apache.commons.io.IOUtils;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.w3c.dom.ls.DOMImplementationLS;
import org.w3c.dom.ls.LSSerializer;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;

/**
 * Created by jon on 28/12/2016.
 */

public class GuideDownloader
{
    private final ResourceManager resourceManager;
    private List<Update> queuedDownloads;

    class Update
    {
        public Update(Element element)
        {
            this.element = element;
        }

        Element element;

        public String getUrl()
        {
            return element.getAttribute("url");
        }

        public String getFilename()
        {
            return element.getAttribute("filename");
        }

        public String getLastModified()
        {
            return element.getAttribute("lastModified");
        }

    }

    class Updates
    {
        Document document;

        Updates(InputStream is)
        {
            if (is != null)
            {
                try
                {
                    document = getDocumentBuilder().parse(is);
                    return;
                }
                catch (Exception e)
                {
                    e.printStackTrace();
                }
            }
            // create default empty
            document = getDocumentBuilder().newDocument();
            document.appendChild( document.createElement("updates") );
        }

        private DocumentBuilder getDocumentBuilder()
        {
            try
            {
                return DocumentBuilderFactory.newInstance().newDocumentBuilder();
            }
            catch (ParserConfigurationException e)
            {
                e.printStackTrace();
            }
            return null;
        }


        Long getMaxLastMod()
        {
            String val = document.getDocumentElement().getAttribute("maxLastMod");
            if (val != null && val.length() > 0)
            {
                return Long.parseLong(val);
            }
            return null;
        }

        void setMaxLastMod(Long lastMod)
        {
            document.getDocumentElement().setAttribute("maxLastMod", lastMod.toString());
        }

        List<Update> getUpdates()
        {
            NodeList nl =  document.getDocumentElement().getElementsByTagName("update");
            List<Update> ret = new ArrayList<>(nl.getLength());

            for (int i=0; i<nl.getLength(); i++)
            {
                Node n = nl.item(i);
                if (n instanceof Element)
                {
                    ret.add( new Update((Element)n) );
                }
            }
            return ret;
        }

        public void save()
        {
            File file = GuideDownloader.this.getUpdatesFilePath();

            try
            {
                Transformer transformer = TransformerFactory.newInstance().newTransformer();
                StreamResult result = new StreamResult(file);
                DOMSource source = new DOMSource(document);
                transformer.transform(source, result);
            }
            catch (Exception e)
            {
                Log.e("GuideDownloader","Unexpected error saving updates.xml", e);
            }
        }

        public void addUpdate(Update update)
        {
            Element newOne = this.document.createElement("update");
            newOne.setAttribute("url", update.getUrl());
            newOne.setAttribute("filename", update.getFilename());
            newOne.setAttribute("lastModified", update.getLastModified());
            document.getDocumentElement().appendChild(newOne);

        }
    }

    static String BASE_URL = "http://www.thesarvo.com/confluence";
    static String  SYNC_URL = BASE_URL + "/plugins/servlet/guide/sync/";

    long since = 0;

    File directory = null;

    int completedOps = 0;
    int totalOps = 0;

    Executor queue = Executors.newSingleThreadExecutor();

    //var taskToUpdate = Dictionary<Int, Update>()

    Updates updates = new Updates(null);

    boolean isSyncing()
    {
        return completedOps < totalOps && totalOps > 0;
    }

    public GuideDownloader(File directory, ResourceManager resourceManager)
    {
        this.directory = directory;
        this.resourceManager = resourceManager;

        queue.execute( () ->
        {
            // load the existing file

            File updateFile = getUpdatesFilePath();
            if (updateFile.exists())
            {
                try
                {
                    this.updates = new Updates( new FileInputStream(updateFile));
                }
                catch (FileNotFoundException e)
                {
                    e.printStackTrace();
                }
            }

            // check if we have a newer resource updates.xml than our local one
            maybeCopyResourceUpdatesXml();

            queue.execute(this::startSync);
        });

    }

    private void maybeCopyResourceUpdatesXml()
    {
        InputStream resourceUpdatesXml = resourceManager.getDataAsset("updates.xml");
        if (resourceUpdatesXml != null)
        {
            Updates resourceUpdates = new Updates(resourceUpdatesXml);
            Long lastMod = resourceUpdates.getMaxLastMod();
            if (lastMod != null)
            {
                if (updates.getMaxLastMod() == null || updates.getMaxLastMod() < lastMod)
                {
                    updates.setMaxLastMod(lastMod);
                    updates.save();
                }
            }
        }
    }

    File getUpdatesFilePath()
    {
        return getFinalPath("updates.xml");
    }

    private File getFinalPath(String filename)
    {
        return new File(directory, filename).getAbsoluteFile();
    }

    private void startSync()
    {
        if (isSyncing())
        {
            Log.d("GuideDownloader", "Already syncing, not starting again");
            return;
        }

        try
        {
            Long s = this.updates.getMaxLastMod();
            if (s == null)
                s = 0L;

            this.completedOps = 0;
            this.totalOps = 1;

            String surl = SYNC_URL + s.toString();
            URL url = new URL(surl);

            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            InputStream in = new BufferedInputStream(conn.getInputStream());
            Updates newUpdates = new Updates(in);
            processNewUpdates(newUpdates);

            this.queuedDownloads = this.updates.getUpdates();
            this.totalOps += this.queuedDownloads.size();
            queue.execute(this::downloadFirst);
        }
        catch (Throwable t)
        {
            Log.e("GuideDownloader", "Error getting updates list", t);

        }
        completedOps++;
    }

    /**
     * Attempt to download the head of the queue
     */
    private void downloadFirst()
    {
        if (queuedDownloads == null || queuedDownloads.isEmpty())
            return;

        // pop the top of the queue
        Update u = queuedDownloads.get(0);
        queuedDownloads.remove(0);
        File finalPath = getFinalPath(u.getFilename());
        String surl = u.getUrl();


        try
        {
            downloadUrl(surl, finalPath);

            // remove from our list
            u.element.getParentNode().removeChild(u.element);
            updates.save();
        }
        catch (Throwable t)
        {
            Log.e("GuideDownloader", "Error downloading file: " + u.getUrl(), t);
        }
        completedOps++;

        if (queuedDownloads.size() > 0)
        {
            queue.execute(this::downloadFirst);
        }
    }

    private void downloadUrl(String surl, File finalPath) throws IOException
    {
        URL url = new URL(surl);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();

        File tempFile = new File( GuideApplication.get().getCacheDir(), UUID.randomUUID().toString() );

        InputStream in = new BufferedInputStream(conn.getInputStream());
        FileUtils.copyInputStreamToFile(in, tempFile);
        IOUtils.closeQuietly(in);
        conn.disconnect();

        // atomically move into place

        if (finalPath.exists())
            finalPath.delete();
        tempFile.renameTo(finalPath);
    }

    private void processNewUpdates(Updates newUpdates)
    {
        List<Update> existing = new ArrayList<>(this.updates.getUpdates());
        for (Update update : newUpdates.getUpdates())
        {
            String f = update.getFilename();
            if (f != null)
            {
                if (!Iterables.any( existing, u ->  f.equals(u.getFilename()) ))
                {
                    // no match, add it
                    this.updates.addUpdate(update);
                }
            }

        }
        this.updates.save();
    }

}

File: thesarvo-android/guide/src/main/java/com/thesarvo/guide/GuideDownloader.kt
package com.thesarvo.guide


import android.util.Log
import com.google.common.collect.Iterables
import org.apache.commons.io.FileUtils
import org.apache.commons.io.IOUtils
import org.w3c.dom.Document
import org.w3c.dom.Element
import java.io.*
import java.net.HttpURLConnection
import java.net.URL
import java.util.*
import java.util.concurrent.Executor
import java.util.concurrent.Executors
import javax.xml.parsers.DocumentBuilder
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

/**
 * Created by jon on 28/12/2016.
 */

class GuideDownloader(directory: File, private val resourceManager: ResourceManager)
{
    private var queuedDownloads: MutableList<Update>? = null

    internal var since: Long = 0

    internal var directory: File? = null

    internal var completedOps = 0
    internal var totalOps = 0


    internal var queue: Executor = Executors.newSingleThreadExecutor()

    internal var updates = Updates(null)

    internal val isSyncing: Boolean
        get() = completedOps < totalOps && totalOps > 0

    internal val updatesFilePath: File
        get() = getFinalPath("updates.xml")

    internal inner class Update(var element: Element)
    {

        val url: String
            get() = element.getAttribute("url")

        val filename: String?
            get() = element.getAttribute("filename")

        val lastModified: String
            get() = element.getAttribute("lastModified")

    }

    internal inner class Updates(`stream`: InputStream?)
    {
        lateinit var document: Document

        private val documentBuilder: DocumentBuilder?
            get()
            {
                try
                {
                    return DocumentBuilderFactory.newInstance().newDocumentBuilder()
                }
                catch (e: ParserConfigurationException)
                {
                    e.printStackTrace()
                }

                return null
            }


        var maxLastMod: Long?
            get()
            {
                val `val` = document.documentElement.getAttribute("maxLastMod")
                return if (`val` != null && `val`.length > 0)
                {
                    java.lang.Long.parseLong(`val`)
                }
                else null
            }
            set(lastMod) = document.documentElement.setAttribute("maxLastMod", lastMod!!.toString())

        var indexHash: String?
            get()
            {
                return document.documentElement.getAttribute("indexHash")
            }
            set(newVal) = document.documentElement.setAttribute("indexHash", newVal)


        val updates: MutableList<Update>
            get()
            {
                val nl = document.documentElement.getElementsByTagName("update")
                val ret = ArrayList<Update>(nl.length)

                for (i in 0 until nl.length)
                {
                    val n = nl.item(i)
                    if (n is Element)
                    {
                        ret.add(Update(n))
                    }
                }
                return ret
            }

        init
        {
            if (`stream` != null)
            {
                try
                {
                    document = documentBuilder!!.parse(`stream`)


                }
                catch (e: Exception)
                {
                    e.printStackTrace()
                }
                finally
                {
                    IOUtils.closeQuietly(stream)
                }


            }
            else
            {
                // create default empty

                document = documentBuilder!!.newDocument()
                document.appendChild(document.createElement("updates"))
            }
        }

        fun save()
        {
            val file = this@GuideDownloader.updatesFilePath

            try
            {
                val transformer = TransformerFactory.newInstance().newTransformer()
                val result = StreamResult(file)
                val source = DOMSource(document)
                transformer.transform(source, result)
            }
            catch (e: Exception)
            {
                Log.e("GuideDownloader", "Unexpected error saving updates.xml", e)
            }

        }

        fun addUpdate(update: Update)
        {
            val newOne = this.document.createElement("update")
            newOne.setAttribute("url", update.url)
            newOne.setAttribute("filename", update.filename)
            newOne.setAttribute("lastModified", update.lastModified)
            document.documentElement.appendChild(newOne)

        }
    }

    init
    {
        this.directory = directory

        queue.execute {
            // load the existing file

            val updateFile = updatesFilePath
            if (updateFile.exists())
            {
                try
                {
                    this.updates = Updates(FileInputStream(updateFile))
                }
                catch (e: FileNotFoundException)
                {
                    e.printStackTrace()
                }

            }

            // check if we have a newer resource updates.xml than our local one
            maybeCopyResourceUpdatesXml()

            queue.execute { this.startSync() }
        }

    }

    private fun maybeCopyResourceUpdatesXml()
    {
        val resourceUpdatesXml = resourceManager.getDataAsset("updates.xml")
        if (resourceUpdatesXml != null)
        {
            val resourceUpdates = Updates(resourceUpdatesXml)
            val lastMod = resourceUpdates.maxLastMod
            if (lastMod != null)
            {
                if (updates.maxLastMod == null || updates.maxLastMod!! < lastMod)
                {
                    updates.maxLastMod = lastMod
                    updates.save()
                }
            }
        }
    }

    private fun getFinalPath(filename: String?): File
    {
        return File(directory, filename!!).absoluteFile
    }

    private fun startSync()
    {
        if (isSyncing)
        {
            Log.d("GuideDownloader", "Already syncing, not starting again")
            return
        }

        try
        {


            this.completedOps = 0
            this.totalOps = 1
            getUpdatesList()

            updateProgress(null)
            queue.execute { this.downloadFirst() }
        }
        catch (t: Throwable)
        {
            Log.e("GuideDownloader", "Error getting updates list", t)

        }

        incrementCompleted()
    }

    private fun getUpdatesList()
    {
        updateProgress("Getting updates")

        var s = this.updates.maxLastMod
        if (s == null)
            s = 0L

        val surl = SYNC_URL + s.toString()
        val url = URL(surl)

        val conn = url.openConnection() as HttpURLConnection
        val `in` = BufferedInputStream(conn.inputStream)
        val newUpdates = Updates(`in`)
        processNewUpdates(newUpdates)

        this.queuedDownloads = this.updates.updates
        this.totalOps += this.queuedDownloads?.size ?: 0
    }



    private fun incrementCompleted()
    {
        completedOps++
        updateProgress(null)
    }

    private fun updateProgress(text: String?)
    {
        var txt = text
        if (txt == null)
        {
            txt = """Updated ${completedOps} of $totalOps"""
        }

        val mainActivity = MainActivity.get()
        mainActivity?.setProgress(completedOps.toLong(), totalOps.toLong(), txt)
    }


    /**
     * Attempt to download the head of the queue
     */
    private fun downloadFirst()
    {
        val queuedDownloads = this.queuedDownloads
        if (queuedDownloads == null || queuedDownloads.isEmpty())
            return

        // pop the top of the queue
        val u = queuedDownloads[0]
        queuedDownloads.removeAt(0)
        val finalPath = getFinalPath(u.filename)
        val surl = u.url


        try
        {
            downloadUrl(surl, finalPath)

            // remove from our list
            u.element.parentNode.removeChild(u.element)
            updates.save()

            // queue an index task
            if (u.filename?.endsWith(".xml") == true)
            {
                val viewId = "guide." + u.filename!!.replace(".xml", "")
                val task = SearchIndexTask(GuideApplication.get(), viewId)

                queue.execute(task)
            }
        }
        catch (t: Throwable)
        {
            Log.e("GuideDownloader", "Error downloading file: " + u.url, t)
        }

        incrementCompleted()

        if (queuedDownloads.size > 0)
        {
            queue.execute { this.downloadFirst() }
        }
    }

    private fun incrementTotal()
    {
        totalOps++
        updateProgress(null)
    }

    @Throws(IOException::class)
    private fun downloadUrl(surl: String, finalPath: File)
    {
        val url = URL(surl)

        val conn = url.openConnection() as HttpURLConnection

        val tempFile = File(GuideApplication.get().cacheDir, UUID.randomUUID().toString())

        val stream = BufferedInputStream(conn.inputStream)
        stream.use {
            FileUtils.copyInputStreamToFile(stream, tempFile)
        }
        conn.disconnect()

        // atomically move into place

        if (finalPath.exists())
            finalPath.delete()
        tempFile.renameTo(finalPath)
    }

    private fun processNewUpdates(newUpdates: Updates)
    {
        val existing = ArrayList(this.updates.updates)
        for (update in newUpdates.updates)
        {
            val f = update.filename
            if (f != null)
            {
                if (!Iterables.any(existing) { u -> f == u?.filename })
                {
                    // no match, add it
                    this.updates.addUpdate(update)
                }
            }

        }
        this.updates.maxLastMod = newUpdates.maxLastMod
        this.updates.save()

        maybeDownloadIndexJson(newUpdates)
    }

    private fun maybeDownloadIndexJson(newUpdates: Updates)
    {
        val newHash = newUpdates.indexHash
        val existingHash = updates.indexHash
        if (newHash != null
                && newHash.isNotEmpty()
                && existingHash != newHash)
        {
            incrementTotal()

            val finalPath = getFinalPath("index.json")

            try
            {
                downloadUrl(INDEX_URL, finalPath)
                updates.indexHash = newHash
                updates.save()
            }
            catch (t: Throwable)
            {
                Log.e("GuideDownloader", "Error downloading index file", t)
            }
            finally
            {
                incrementCompleted()
            }
        }

    }

    companion object
    {

        internal var BASE_URL = "http://www.thesarvo.com/confluence"
        internal var SYNC_URL = "$BASE_URL/plugins/servlet/guide/sync/"
        internal var INDEX_URL = "$BASE_URL/plugins/servlet/guide/index/1"
    }

}
